slipp/src/main/java/slipp/controller/ProfileController.java

```diff
@@ -1,17 +1,19 @@
 package slipp.controller;
 
-import slipp.domain.User;
-import slipp.support.db.DataBase;
 import nextstep.mvc.asis.Controller;
+import slipp.dao.UserDao;
+import slipp.domain.User;
 
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 public class ProfileController implements Controller {
+    private final UserDao userDao = new UserDao();
+
     @Override
     public String execute(HttpServletRequest req, HttpServletResponse resp) throws Exception {
         String userId = req.getParameter("userId");
-        User user = DataBase.findUserById(userId);
+        User user = userDao.findByUserId(userId);
```

Comment:
Optional로 받아서 exception 처리하면 Controller로직이 더 깔끔해질 수 있을 것 같아요.
```suggestion
        User user = userDao.findByUserId(userId).orElseThrow(() -> new NullPointerException("사용자를 찾을 수 없습니다."));  
```

이와 함께, NullPointerException 대신 의미있는 custom exception을 만들어보는 연습을 해보셔도 좋을 것 같습니다.

Reply:
Optional 로 변경해보았습니다..! 

변경하면서 LoginController 로직에서 조금 고민되는 부분이 있었습니다

__Old__
``` java
    public String execute(HttpServletRequest req, HttpServletResponse resp) throws Exception {
        String userId = req.getParameter("userId");
        String password = req.getParameter("password");
        User user = DataBase.findUserById(userId);
        if (user == null) {
            req.setAttribute("loginFailed", true);
            return "/user/login.jsp";
        }
        if (user.matchPassword(password)) {
            HttpSession session = req.getSession();
            session.setAttribute(UserSessionUtils.USER_SESSION_KEY, user);
            return "redirect:/";
        } else {
            req.setAttribute("loginFailed", true);
            return "/user/login.jsp";
        }
    }
```

New
``` java
    public String execute(HttpServletRequest req, HttpServletResponse resp) throws Exception {
        String userId = req.getParameter("userId");
        String password = req.getParameter("password");

        return userDao.findByUserId(userId)
                .filter(user -> user.matchPassword(password))
                .map(user -> login(req, user))
                .orElseGet(() -> failLogin(req));
    }

    private String login(HttpServletRequest req, User user) {
        HttpSession session = req.getSession();
        session.setAttribute(UserSessionUtils.USER_SESSION_KEY, user);
        return "redirect:/";
    }

    private String failLogin(HttpServletRequest req) {
        req.setAttribute("loginFailed", true);
        return "/user/login.jsp";
    }
```

Optional 로 바뀌니 기존의 null 로 미리 끝내는 부분(early exit) 이 조금 문제였습니다.
`Optional.isPresent` 를 사용하고 싶지 않아서 고민하다 보니 New 와 같은 구조가 되었네요

이런식으로 주로 사용할지, 또한 이렇게 구현했을 때 가독성에서 문제가 없을까요?
(저는 early exit 에 조금은 익숙해서요. 보통 New 처럼 해도 코드를 이해하는데 불편함이 없을까.. 궁금합니다.)